source/h1_encoder.c

```diff
@@ -755,6 +750,134 @@ static int s_state_fn_unchunked_body(struct aws_h1_encoder *encoder, struct aws_
+            aws_error_name(aws_last_error()));
+        return AWS_OP_ERR;
+    }
+
+    /* If ANY body data was streamed, then write in chunk prefix and suffix.
+     *
+     * (else no body data streamed, so dst remains untouched. Maybe we've
+     * reached end of stream, maybe user just doesn't have data yet to send) */
+    if (body_sub_buf.len > 0) {
+        encoder->chunk_count++;
+        ENCODER_LOGF(
+            TRACE, encoder, "Sending chunk #%" PRIu64 " with size %zu", encoder->chunk_count, body_sub_buf.len);
+
+        /* Write chunk-prefix: LENGTH-IN-HEX CRLF */
+        char hexbuf[padded_hex_len + 1] = {0};
+        snprintf(hexbuf, sizeof(hexbuf), padded_hex_fmt, body_sub_buf.len);
+        aws_byte_buf_write_from_whole_cursor(dst, aws_byte_cursor_from_c_str(hexbuf));
+        s_write_crlf(dst);
+
+        /* Increment dst->len, since we already copied body in there via sub-buffer */
+
+        dst->len += body_sub_buf.len;
+        AWS_ASSERT(dst->len <= dst->capacity);
+
+        /* Write chunk-suffix: CRLF */
+        s_write_crlf(dst);
+    }
+
+    /* If body stream has ended: switch states.
+     * As an optimization, we only do this check when the stream didn't 100% fill the buffer */
+    if (body_sub_buf.len < body_sub_buf.capacity) {
+        struct aws_stream_status stream_status;
+        if (aws_input_stream_get_status(encoder->message->body, &stream_status) != AWS_OP_SUCCESS) {
+            ENCODER_LOGF(
+                ERROR,
+                encoder,
+                "Failed to query body stream status, error %d (%s)",
+                aws_last_error(),
+                aws_error_name(aws_last_error()));
+            return AWS_OP_ERR;
+        }
+
+        if (stream_status.is_end_of_stream) {
+            encoder->chunk_count++;
+            ENCODER_LOGF(TRACE, encoder, "Sending last chunk #%" PRIu64, encoder->chunk_count);
+            return s_switch_state(encoder, AWS_H1_ENCODER_STATE_CHUNKED_BODY_STREAM_LAST_CHUNK);
+        }
+    }
+
+    /* Remain in state until done streaming body */
+    return AWS_OP_SUCCESS;
+}
+
+/* Note: this state is ONLY used when streaming a body of unknown Content-Length.
+ * It is NOT used when the write_chunk() API is being used. */
+static int s_state_fn_chunked_body_stream_last_chunk(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst) {
+    struct aws_byte_cursor last_chunk = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("0\r\n");
+
+    const size_t space_available = dst->capacity - dst->len;
+    if (space_available < last_chunk.len) {
+        /* Remain in state until there's enough space to write */
+        return AWS_OP_SUCCESS;
+    }
+
+    if (aws_byte_buf_write_from_whole_cursor(dst, last_chunk) == true) {
+        ENCODER_LOG(TRACE, encoder, "Last chunk complete");
+        return s_switch_state(encoder, AWS_H1_ENCODER_STATE_CHUNK_TRAILER);
+    } else {
+        /* Remain in state until there's enough space to write */
+        return AWS_OP_SUCCESS;
+    }
```

Comment:
don't we just checked the space couple line before? should we just assert on the return val?

Reply:
this is a the `CHUNKED_BODY_STREAM_LAST_CHUNK` state, that we go into after aws_input_stream has reported EOF

so like imagine:
-  in the `AWS_H1_ENCODER_STATE_CHUNKED_BODY_STREAM` state, the aws_input_stream nearly fills the aws_io_message, there are only 2 bytes left over. The state checks if the stream is EOF, and it is, so the state machine moves to the next state...
- This is the AWS_H1_ENCODER_STATE_CHUNKED_BODY_STREAM_LAST_CHUNK, which we're looking at. It want to write "0\r\n", it needs 3 bytes, so we `/* Remain in state until there's enough space to write */`
- The h1_encoder state machine sees that the state didn't change, so returns control back to the h1_connection, which sends the aws_io_message and runs the encoder again in some future tick with a fresh aws_io_message...

Reply:
whoops, didn't mean to be redundant with the checks